plugin/pkg/proxy/proxy.go

```diff
@@ -100,80 +100,91 @@ func (k *Currency) GetParameters() map[string]string {
 func (k *Currency) OnRequest(id uint64, payload []byte, hasSURB bool) ([]byte, error) {
 	k.log.Debugf("Handling request %d", id)
 
-	// Send request to HTTP RPC.
+	// Decode request
 	req, err := common.RequestFromJson(payload)
 	if err != nil {
 		k.log.Debug("Failed to decode currency transaction request: (%v)", err)
 		return common.RespondFailure(err), nil
 	}
-	if _, ok := k.rpc[req.Ticker]; !ok {
+
+	// Get supported chain
+	rpc, ok := k.rpc[req.Ticker]
+	if !ok {
 		return nil, common.ErrWrongTicker
 	}
+	c, err := chain.GetChain(req.Ticker)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap command into transactions
+	sendData, err := c.WrapRequest(rpc.Url, req.Command, req.Payload)
+	if err != nil {
+		return nil, err
+	}
 
-	hash, err := k.sendTransaction(req.Ticker, req.Tx)
+	// Send the transactions
+	response, err := k.sendTransaction(rpc, sendData)
 	if err != nil {
 		k.log.Debug("Failed to send currency transaction request: (%v)", err)
 		return common.RespondFailure(err), nil
 	}
-	return common.RespondSuccess("Tx hash: " + hash), nil
+
+	// Unwrap responses
+	result, err := c.UnwrapResponse(req.Command, response)
+	if err != nil {
+		return nil, err
+	}
+
+	return common.RespondSuccess(string(result)), nil
 }
 
 // Halt : Stops the plugin
 func (k *Currency) Halt() {
 
 }
 
-func (k *Currency) sendTransaction(ticker string, txHex string) (string, error) {
+func (k *Currency) sendTransaction(rpc config.RPCMetadata, sendData []chain.HttpData) ([]string, error) {
 	k.log.Debug("sendTransaction")
-
-	// Get supported chain
-	c, err := chain.GetChain(ticker)
-	if err != nil {
-		return "", err
-	}
-	rpc := k.rpc[ticker]
-	// Create a new appropriately marshalled request
-	postRequest, err := c.NewRequest(rpc.Url, txHex)
-	if err != nil {
-		return "", err
-	}
-
-	bodyReader := bytes.NewReader(postRequest.Body)
-
-	// create an http request
-	httpReq, err := http.NewRequest("POST", postRequest.URL, bodyReader)
-	if err != nil {
-		return "", err
-	}
-	httpReq.Close = true
-	httpReq.Header.Set("Content-Type", "application/json")
-	if rpc.User != "" && rpc.Pass != "" {
-		httpReq.SetBasicAuth(rpc.User, rpc.Pass)
-	}
-
-	// send http request
-	client := http.Client{}
-	httpResponse, err := client.Do(httpReq)
-	if err != nil {
-		return "", err
-	}
-	if httpResponse.StatusCode != http.StatusOK {
-		return "", fmt.Errorf("currency RPC error status: %s", httpResponse.Status)
-	}
-	resp := RPCResponse{}
-	bodyBytes, err := ioutil.ReadAll(httpResponse.Body)
-	if err != nil {
-		return "", err
-	}
-	dec := codec.NewDecoderBytes(bodyBytes, &jsonHandle)
-	err = dec.Decode(&resp)
-	if err != nil {
-		return "", err
-	}
-	if resp.Error != nil {
-		return "", errors.New(resp.Error.Message)
-	}
-	return resp.Result, nil
+	var result []string
+
+	for _, send := range sendData {
+		bodyReader := bytes.NewReader(send.Body)
+		httpReq, err := http.NewRequest("POST", send.URL, bodyReader)
```

Comment:
If we add Method, the request should update.

Reply:
Also there is batchRequest in jsonRPC spec, we might need to add that functionality to send one request instead of multiple requests, eg. in ethereum we send three requests to query account/gas information.

Reply:
Agree. I use batch request now.